smoke/LeafDevice/details/Details.cs

```diff
@@ -143,29 +143,56 @@ protected Task InitializeTrustedCertsAsync()
-                deviceClient = DeviceClient.Create(builder.HostName, this.edgeHostName, auth, this.deviceTransportSettings);
-            }
+                Exception savedException = null;
+
+                try
+                {
+                    var builder = IotHubConnectionStringBuilder.Create(this.iothubConnectionString);
+                    DeviceClient deviceClient;
+                    if (this.authType == AuthenticationType.Sas)
+                    {
+                        string leafDeviceConnectionString = $"HostName={builder.HostName};DeviceId={this.deviceId};SharedAccessKey={this.context.Device.Authentication.SymmetricKey.PrimaryKey};GatewayHostName={this.edgeHostName}";
+                        deviceClient = DeviceClient.CreateFromConnectionString(leafDeviceConnectionString, this.deviceTransportSettings);
+                    }
+                    else
+                    {
+                        var auth = new DeviceAuthenticationWithX509Certificate(this.deviceId, this.clientCertificate.Expect(() => new InvalidOperationException("Missing client certificate")));
+                        deviceClient = DeviceClient.Create(builder.HostName, this.edgeHostName, auth, this.deviceTransportSettings);
+                    }
 
-            this.context.DeviceClientInstance = Option.Some(deviceClient);
-            Console.WriteLine("Leaf Device client created.");
+                    this.context.DeviceClientInstance = Option.Some(deviceClient);
+                    Console.WriteLine("Leaf Device client created.");
 
-            var message = new Message(Encoding.ASCII.GetBytes($"Message from Leaf Device. Msg GUID: {this.context.MessageGuid}"));
-            Console.WriteLine($"Trying to send the message to '{this.edgeHostName}'");
+                    var message = new Message(Encoding.ASCII.GetBytes($"Message from Leaf Device. Msg GUID: {this.context.MessageGuid}"));
+                    Console.WriteLine($"Trying to send the message to '{this.edgeHostName}'");
 
-            await deviceClient.SendEventAsync(message);
-            Console.WriteLine("Message Sent.");
-            await deviceClient.SetMethodHandlerAsync("DirectMethod", DirectMethod, null);
-            Console.WriteLine("Direct method callback is set.");
+                    while (!cts.IsCancellationRequested) // Retries are needed as the DeviceClient timeouts are not long enough
+                    {
+                        try
+                        {
+                            await deviceClient.SendEventAsync(message);
+                            Console.WriteLine("Message Sent.");
+                            await deviceClient.SetMethodHandlerAsync("DirectMethod", DirectMethod, null);
+                            Console.WriteLine("Direct method callback is set.");
+                            break;
+                        }
+                        catch (Exception e)
```

Comment:
does it throw a specific exception when the device is registered, but not ready to use? if so, please catch that specific exception instead.

Reply:
I will address this in a PR to master then port back to 109